### PR TITLE
fix/integration_tests: avoid adding peers too aggressively

### PR DESCRIPTION
**Pull request by Fraser999**
This resolves an issue in the test framework where we would add many peers in a short period, and then start dropping the existing ones as the new ones are voted in.  When that happened, a new peer could end up not receiving gossip from any of the peers with which it was initialised, and would hence be stalled as it wouldn't accept messages from any of the new peers.

This commit changes the test framework to only allow the addition and voting for addition of new peers if that would retain a supermajority of "Joined" peers out of "Joined and Joining" ones.  When this is not the case, addition and voting is postponed, not cancelled.

### CODE REVIEW

File: src/dev_utils/network.rs [hunk 1]
```diff
@@ -581,15 +598,21 @@ impl Network {
...
             }
-            NetworkView::Leaving | NetworkView::Left => true,
+            Some(NetworkView::Leaving) | Some(NetworkView::Left) => true,
         }
     }
+
+    fn allow_addition_of_peer(&self) -> bool {
+        let joined_count = self.num_with_network_view(NetworkView::Joined);
+        let joining_count = self.num_with_network_view(NetworkView::Joining);
+        is_more_than_two_thirds(joined_count, joined_count + joining_count + 1)
```
Comment (fizyk20) [inline src/dev_utils/network.rs]: Why `+ 1`? It could use a comment, I think.

File: tests/integration_tests.rs [hunk 2]
```diff
@@ -219,7 +219,7 @@ fn random_schedule_no_delays() {
 fn add_many_peers() {
     let mut env = Environment::new(SEED);
     let options = ScheduleOptions {
-        genesis_size: 2,
+        genesis_size: 3,
```
Comment (fizyk20) [inline tests/integration_tests.rs]: Why increase the starting size? Shouldn't the network still work starting with just 2 nodes?

File: tests/integration_tests.rs [hunk 3]
```diff
@@ -259,7 +259,7 @@ fn add_few_peers_and_vote() {
 fn add_many_peers_and_vote() {
     let mut env = Environment::new(SEED);
     let options = ScheduleOptions {
-        genesis_size: 2,
+        genesis_size: 3,
```
Comment (fizyk20) [inline tests/integration_tests.rs]: Same as above.

File: tests/integration_tests.rs (hunk 2 above)
Comment (Fraser999) [reply tests/integration_tests.rs]: Yes, it should, but the new `allow_addition_of_peer()` prohibits adding a peer if it results in less than a supermajority of joined peers, which is the case if we try to add a single peer to a network of two.

I considered putting `|| (joined_count == 2 && joining_count == 0)` in `allow_addition_of_peer()` to support this specific case, but thought starting with three peers was cleaner.  I'm happy to change that though, or is there a better option you can see?

File: src/dev_utils/network.rs [hunk 4]
```diff
@@ -561,17 +564,31 @@ impl Network {
...
+                    ParsecObservation::Remove { ref peer_id, .. } => {
+                        if self.allow_removal_of_peer(&peer_id) {
+                            (*self.peer_mut(&peer_id)).mark_network_view_as_leaving();
+                        } else {
+                            return Ok(false);
+                        }
+                    }
+                    ParsecObservation::Add { ref peer_id, .. } => {
+                        // If the peer can't be added yet, postpone voting for its addition.
+                        if !self.peers.contains_key(peer_id) {
```
Comment (maqi) [inline src/dev_utils/network.rs]: this code seems only checks the existence, which doens't match the comment above?
or do I miss something?

Comment (Fraser999) [reply src/dev_utils/network.rs]: When we hit `ScheduleEvent::AddPeer` we now check to see if we're able to add a new peer, and if so, it's added to `self.peers`.  I'm using this mechanism (the existence of the requested new peer in `self.peers`) as an indicator to decide whether any other peer should cast its vote for the new peer yet or not.

Comment (maqi) [reply src/dev_utils/network.rs]: Then can we make the comment to be more explicit?
The comment is bit confusing without knowing the context of the mechanism you explained.

File: tests/integration_tests.rs (hunk 2 above)
Comment (fizyk20) [reply tests/integration_tests.rs]: I'm worried this might indicate a flaw in the logic behind `allow_addition_of_peer()`, but it shouldn't be a big problem (at worst, we exclude some edge cases from being generated), so I'm fine with leaving it as is.
